syncer/client.go

```diff
@@ -0,0 +1,277 @@
+
+	status, err := clt.GetStatus(timeoutCtx, &emptypb.Empty{})
+	if err != nil {
+		return nil, err
+	}
+
+	return &NoForkPeer{
+		ID:       peerID,
+		Number:   status.Number,
+		Distance: m.network.GetPeerDistance(peerID),
+	}, nil
+}
+
+func (m *syncPeerClient) GetConnectedPeerStatuses() []*NoForkPeer {
+	var (
+		ps            = m.network.Peers()
+		syncPeers     = make([]*NoForkPeer, 0, len(ps))
+		syncPeersLock = sync.Mutex{}
+		wg            sync.WaitGroup
+	)
+
+	for _, p := range ps {
+		p := p
+
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			peerID := p.Info.ID
+
+			status, err := m.GetPeerStatus(peerID)
+			if err != nil {
+				m.logger.Warn("failed to get status from a peer, skip", "id", peerID, "err", err)
+			}
+
+			syncPeersLock.Lock()
+
+			syncPeers = append(syncPeers, status)
+
+			syncPeersLock.Unlock()
+		}()
+	}
+
+	wg.Wait()
+
+	return syncPeers
+}
+
+func (m *syncPeerClient) GetPeerStatusUpdateCh() <-chan *NoForkPeer {
+	return m.peerStatusUpdateCh
+}
+
+func (m *syncPeerClient) GetPeerConnectionUpdateEventCh() <-chan *event.PeerEvent {
+	return m.peerConnectionUpdateCh
+}
+
+func (m *syncPeerClient) startGossip() error {
+	topic, err := m.network.NewTopic(statusTopicName, &proto.SyncPeerStatus{})
+	if err != nil {
+		return err
+	}
+
+	if err := topic.Subscribe(m.handleStatusUpdate); err != nil {
+		return fmt.Errorf("unable to subscribe to gossip topic, %w", err)
+	}
+
+	m.topic = topic
+
+	return nil
+}
+
+func (m *syncPeerClient) handleStatusUpdate(obj interface{}, from peer.ID) {
+	status, ok := obj.(*proto.SyncPeerStatus)
+	if !ok {
+		m.logger.Error("failed to cast gossiped message to txn")
+
+		return
+	}
+
+	if !m.network.IsConnected(from) {
+		if m.id != from.String() {
+			m.logger.Warn("received status from non-connected peer, ignore", "id", from)
+		}
+
+		return
+	}
+
+	m.peerStatusUpdateCh <- &NoForkPeer{
+		ID:       from,
+		Number:   status.Number,
+		Distance: m.network.GetPeerDistance(from),
+	}
+}
+
+func (m *syncPeerClient) startNewBlockProcess() {
+	m.subscription = m.blockchain.SubscribeEvents()
+
+	for event := range m.subscription.GetEventCh() {
+		if event.Type == blockchain.EventHead && len(event.NewChain) > 0 {
+			latest := m.blockchain.Header()
+			if latest == nil {
+				return
+			}
+
+			// Publish status
+			if err := m.topic.Publish(&proto.SyncPeerStatus{
+				Number: latest.Number,
+			}); err != nil {
+				m.logger.Warn("failed to publish status", "err", err)
+			}
+		}
+	}
+}
+
+func (m *syncPeerClient) startPeerEventProcess() {
+	peerEventCh, err := m.network.SubscribeCh()
+	if err != nil {
+		m.logger.Error("failed to subscribe", "err", err)
+
+		return
+	}
+
+	for e := range peerEventCh {
+		switch e.Type {
+		case event.PeerConnected, event.PeerDisconnected:
+			m.peerConnectionUpdateCh <- e
+		}
+	}
+}
+
+func (m *syncPeerClient) CloseStream(peerID peer.ID) error {
+	return m.network.CloseProtocolStream(SyncerProto, peerID)
+}
+
+func (m *syncPeerClient) GetBlocks(
```

Comment:
What do you think about naming this `GetBlocksFromPeer` to avoid confusion with the gRPC method `GetBlocks`?

Reply:
This is a `SyncPeerClient`. So `GetBlocks` means to get blocks from peer 😂

Reply:
Yes, but this is not the server implementation of the gRPC method, but a clientside method that exists independently.
Anybody who's searching for the implementation of GetBlocks (the gRPC method) will stumble upon this one

Reply:
Yes, but this struct is responsible for only interacting with peer...

Reply:
I like this discussion hahaha both of you have a point. Personally, I am ok leaving it as it is. But I really like the way of thinking about how people will search for this.

Reply:
@Kourin1996 

You defined this method **as a part of** `SyncPeerClient`, where you pass in the **peer you want to get blocks from**, as part of the arguments. It makes sense to be called `GetBlocks` on the server side of this interaction (which it is), not so much on the SyncPeer**Client** side

Reply:
This is different things. What is the special reason to use different name on server side and client side?

Reply:
As you said, this method take peerID to specify the peer client fetch data from

Reply:
@Aleksao998 cc